pkg/restore/prioritize_group_version.go

```diff
@@ -0,0 +1,346 @@
+func userResourceGroupVersionPriorities(ctx *restoreContext, cm *corev1.ConfigMap) map[string]metav1.APIGroup {
+	if cm == nil {
+		ctx.log.Debugf("No enableapigroupversion config map found in velero namespace. Using pre-defined priorities.")
+		return nil
+	}
+
+	priorities := parseUserPriorities(cm.Data["restoreResourcesVersionPriority"])
+	if len(priorities) == 0 {
+		ctx.log.Debugf("No valid user version priorities found in enableapigroupversion config map. Using pre-defined priorities.")
+		return nil
+	}
+
+	return priorities
+}
+
+func userPriorityConfigMap() (*corev1.ConfigMap, error) {
+	cfg, err := client.LoadConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "reading client config file")
+	}
+
+	fc := client.NewFactory("APIGroupVersionsRestore", cfg)
+
+	kc, err := fc.KubeClient()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting Kube client")
+	}
+
+	cm, err := kc.CoreV1().ConfigMaps("velero").Get(
+		context.Background(),
+		"enableapigroupversions",
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+
+		return nil, errors.Wrap(err, "getting enableapigroupversions config map from velero namespace")
+	}
+
+	return cm, nil
+}
+
+func parseUserPriorities(prioritiesData string) map[string]metav1.APIGroup {
+	ups := make(map[string]metav1.APIGroup)
+
+	// The user priorities will be in a string of the form
+	// rockbands.music.example.io=v2beta1,v2beta2\n
+	// orchestras.music.example.io=v2,v3alpha1\n
+	// subscriptions.operators.coreos.com=v2,v1
+
+	lines := strings.Split(prioritiesData, "\n")
+
+	for _, l := range lines {
+		if !isUserPriorityValid(l) {
+			continue
+		}
+
+		rgvs := strings.SplitN(l, "=", 2)
+		rg := rgvs[0]   // rockbands.music.example.io
+		vers := rgvs[1] // v2beta1,v2beta2
+
+		vs := strings.Split(vers, ",")
+
+		ups[rg] = metav1.APIGroup{
+			Versions: versionsToGroupVersionForDiscovery(vs),
+		}
+	}
+
+	return ups
+}
+
+func isUserPriorityValid(line string) bool {
+	// Line must have one and only one equal sign
+	if strings.Count(line, "=") != 1 {
+		return false
+	}
+
+	// Line must contain at least one character before and after equal sign
+	pair := strings.Split(line, "=")
+	if len(pair[0]) < 1 || len(pair[1]) < 1 {
+		return false
+	}
+
+	// Line must not contain any spaces
+	if strings.Count(line, " ") > 0 {
+		return false
+	}
+
+	return true
+}
+
+// versionsToGroupVersionForDiscovery converts version strings into a Kubernetes format
+// for group versions.
+func versionsToGroupVersionForDiscovery(vs []string) []metav1.GroupVersionForDiscovery {
+	gvs := make([]metav1.GroupVersionForDiscovery, len(vs))
+
+	for i, v := range vs {
+		gvs[i] = metav1.GroupVersionForDiscovery{
+			Version: v,
+		}
+	}
+
+	return gvs
+}
+
+// findTargetGroup looks for a particular group in from a list of groups in the
+// target cluster.
+func findTargetGroup(tgvs []metav1.APIGroup, group string) metav1.APIGroup {
```

Comment:
I wonder if we can make this function with a more generic name - not only pinning its use for "target".  looking up a group string in a metav1.APIGroup can be used in other circumstances, or no?

Reply:
Yes, I can make it more generic, but it cannot be used for the Source API Groups and User API Groups. The reason is because Target API Groups are a slice of API Groups while the other two are maps where resource groups are the keys and the API Groups are the values. But, not being able to accept both maps and keys is not a problem as developers can read function signatures.

I'll make it more generic and call the function `containsAPIGroup(...)`.